source/common/secret/sds_api.cc

```diff
@@ -0,0 +1,87 @@
+#include "common/secret/sds_api.h"
+
+#include <unordered_map>
+
+#include "envoy/api/v2/auth/cert.pb.validate.h"
+
+#include "common/config/resources.h"
+#include "common/config/subscription_factory.h"
+#include "common/protobuf/utility.h"
+#include "common/ssl/tls_certificate_config_impl.h"
+
+namespace Envoy {
+namespace Secret {
+
+SdsApi::SdsApi(const LocalInfo::LocalInfo& local_info, Event::Dispatcher& dispatcher,
+               Runtime::RandomGenerator& random, Stats::Store& stats,
+               Upstream::ClusterManager& cluster_manager, Init::Manager& init_manager,
+               const envoy::api::v2::core::ConfigSource& sds_config, std::string sds_config_name,
+               std::function<void()> destructor_cb)
```

Comment:
This is better than before, thanks. I still think it can be cleaner if instead of having `SdsApi` owning the destructor CB, we have the owner of `SdsApi` own that object. It's pretty weird threading a destructor CB into an object when the lifetime issues are really those of the owning object.

Reply:
Thanks for pointing this out. My concern is that SdsApi is held by a shared_ptr (ContextConfigImpland::tls_certficate_provider_) and could be owned by multiple objects (e.g. multiple listeners or clusters). The last owner needs to destroy SdsApi and ~SdsApi() invokes this cleanup function. I am not clear if there is a better place to own the Cleanup object other than SdsApi.

Reply:
@htuch SdsApi is a shared pointer that could be owned by multiple objects as @JimmyCYJ pointed out, the structure here is similar to what we have in RDS (`RdsRouteConfigSubscription`). With the destructor_cb it doesn't need to be a friend class between SecretManagerImpl and SdsApi.

Reply:
Yeah, OK, I still think you could wrap `SdsApi` in a `SdsApiWithManagedLifetime` object to get the same effect and separate concerns here, but I guess we have precedent for doing it this way.